Fixed validation for Workshops and CompetitiveEvents

HlibBondarev opened the pull request:
1) Fixed validation of Workshop properties.
2) Fixed validation of CompetitiveEvent properties. 
3) Normalize Conditional fields (CompetitiveSelectionDescription, PreferentialTermsOfParticipation, Price, MinAge and MaxAge) in Create(), Update() and UpdateWorkshop() methods of WorkshopDraftService class for case when {CompetitiveSelection, AreThereBenefits, IsPaid} = false and NoAgeRestrictions = true. 
4) Normalize Conditional fields (CompetitiveSelectionDescription, Benefits and Price) in Create(), Update() and UpdateWorkshop() methods of CompetitiveEventDraftService class for case when {CompetitiveSelection, AreThereBenefits, IsPaid} = false. 
5) Fixed generators for Workshop dtos.
6) Added tests for validation of WorkshopV2Dto and WorkshopV2CreateRequestDto.
7) Added migration — 20251115101013_FixValidationForWorkshopAndWorkshopDescriptionItem.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Enforced image rules (cover exclusivity, required image source, max total images) and required price/pay-rate for paid items.

* **Bug Fixes**
  * Fixed date/time comparison edge cases and tightened age/seat validation to prevent inconsistent submissions.
  * Clarified price range validation.

* **Improvements**
  * Stronger text/keyword/benefits validation (per-item and total limits, allowed characters, duplicate detection), clearer validation messages, and normalized conditional fields.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

HlibBondarev committed:
1) Fixed validation of Workshop properties.
2) Fixed validation of CompetitiveEvent properties.
3) Normalize Conditional fields (CompetitiveSelectionDescription, PreferentialTermsOfParticipation, Price, MinAge and MaxAge) in Create(), Update() and UpdateWorkshop() methods of WorkshopDraftService class for case when {CompetitiveSelection, AreThereBenefits, IsPaid} = false and NoAgeRestrictions = true.
4) Normalize Conditional fields (CompetitiveSelectionDescription, Benefits and Price) in Create(), Update() and UpdateWorkshop() methods of CompetitiveEventDraftService class for case when {CompetitiveSelection, AreThereBenefits, IsPaid}.
5) Fixed generators for Workshop dtos.
6) Added tests for validation of WorkshopV2CreateRequestDto and WorkshopV2CreateRequestDto.

HlibBondarev committed:
Fixed the problem mentioned by SonarCloud.

HlibBondarev committed:
Fixed the problem mentioned by SonarCloud.

HlibBondarev committed:
Fixed the problem mentioned by SonarCloud.

HlibBondarev committed:
Fixed validation of properties (registrationEndTime, registrationStartTime and scheduledStartTime) for CompetitiveEvent.

HlibBondarev commented on OutOfSchool/OutOfSchool.BusinessLogic/Models/CompetitiveEvent/TempSave/CompetitiveEventDescriptionDto.cs:
```diff
@@ -71,5 +75,11 @@ public override IEnumerable<ValidationResult> Validate(ValidationContext validat
         // Run validations from CompetitiveEventAboutDto
         foreach (var error in base.Validate(validationContext))
             yield return error;
+
+        // validate Price when IsPaid is true
+        if (IsPaid && (!Price.HasValue || Price < 0.01M))
+        {
+            yield return new ValidationResult("Price must be specified and must be in the range from 0.01 to 100000.00 when the competitive event is paid.", [nameof(Price)]);
+        }
```
No, Price should be 0.00 when IsPaid = false, but if IsPaid=true Price should be in a range from 0.01 to 100 000. Though error message  "Field value should be in a range from 0 to 100 000" is correct

HlibBondarev commented on OutOfSchool/OutOfSchool.BusinessLogic/Models/CompetitiveEvent/CompetitiveEventBaseDto.cs:
```diff
@@ -154,6 +155,12 @@ public virtual IEnumerable<ValidationResult> Validate(ValidationContext validati
             yield return new ValidationResult("Minimum age should be less than Maximum age", [nameof(MinimumAge), nameof(MaximumAge)]);
         }
+
+        // validate Price when IsPaid is true
+        if (IsPaid && (!Price.HasValue || Price < 0.01M))
+        {
+            yield return new ValidationResult("Price must be specified and must be in the range from 0.01 to 100000.00 when the competitive event is paid.", [nameof(Price)]);
+        }
```
No, Price should be 0.00 when IsPaid = false, but if IsPaid=true Price should be in a range from 0.01 to 100 000. Though error message "Field value should be in a range from 0 to 100 000" is correct

HlibBondarev commented on OutOfSchool/OutOfSchool.BusinessLogic/Models/CompetitiveEventDraft/ModeratorCompetitiveEventDraftEditDto.cs:
```diff
@@ -16,7 +16,7 @@ public class ModeratorCompetitiveEventDraftEditDto
     public string ShortTitle { get; set; }    
 
-    [MaxLength(Constants.EnrollmentProcedureDescription)]
+    [MaxLength(Constants.MaxLengthOfEnrollmentProcedureDescriptionForWorkshop)]
```
Ok

HlibBondarev commented on OutOfSchool/OutOfSchool.BusinessLogic/Models/CompetitiveEvent/TempSave/CompetitiveEventDescriptionDto.cs:
```diff
@@ -37,17 +37,20 @@ public class CompetitiveEventDescriptionDto : CompetitiveEventAboutDto
     [MustContain(RequiredCharacterType.AnyLetter)]
     public string VenueName { get; set; }
 
+    [Required(ErrorMessage = "Information about the selection is required")]
```
The DescriptionOfTheEnrollmentProcedure property should be [Required] according to the validation rules. I added the [Required] attribute to this property in ModeratorCompetitiveEventDraftEditDto. But CompetitiveEventInfoDto gets data from the database, so it doesn't need the [Required] attribute.

HlibBondarev commented on OutOfSchool/OutOfSchool.BusinessLogic/Models/Exported/CompetitiveEvents/CompetitiveEventInfoDto.cs:
```diff
@@ -49,7 +52,7 @@ public class CompetitiveEventInfoDto : CompetitiveEventInfoBaseDto, IExternalRat
 
-    [MaxLength(Constants.EnrollmentProcedureDescription)]
+    [MaxLength(Constants.MaxLengthOfEnrollmentProcedureDescriptionForWorkshop)]
     public string DescriptionOfTheEnrollmentProcedure { get; set; } = string.Empty;
```
CompetitiveEventInfoDto gets data from the database, so it doesn't need the [MaxLength] attribute. So I've removed it complitly

HlibBondarev commented on OutOfSchool/Tests/OutOfSchool.Tests.Common/TestDataGenerators/WorkshopBaseDtoGenerator.cs:
```diff
@@ -15,14 +17,19 @@ public static class WorkshopBaseDtoGenerator
+        .RuleFor(x => x.AgeComposition, AgeComposition.SameAge)
         .RuleFor(x => x.Title, f => f.Company.CompanyName())
+        .RuleFor(x => x.ShortTitle, f => f.Company.CompanyName())
+        .RuleFor(x => x.NoAgeRestrictions, false)
         .RuleFor(x => x.MinAge, f => f.Random.Number(1, 15))
         .RuleFor(x => x.MaxAge, f => f.Random.Number(16, 18))
-        .RuleFor(x => x.DateTimeRanges, f => DateTimeRangeDtoGenerator.Generate(4))
+        .RuleFor(x => x.IsPaid, true)
         .RuleFor(x => x.Price, f => f.Random.Decimal(0, 10000))
-        .RuleFor(x => x.PayRate, f => f.PickRandom<PayRateType>())
+        .RuleFor(x => x.PayRate, f => f.PickRandom(PayRateType.Hour, PayRateType.Day, PayRateType.Month, PayRateType.Year, PayRateType.Course))
         .RuleFor(x => x.AvailableSeats, f => f.Random.UInt(5, 50))
-        .RuleFor(x => x.CompetitiveSelection, f => f.Random.Bool())
+        .RuleFor(x => x.AreThereBenefits, true)
+        .RuleFor(x => x.PreferentialTermsOfParticipation, f => f.Lorem.Sentence())
+        .RuleFor(x => x.CompetitiveSelection, true)
```
ok

HlibBondarev commented on OutOfSchool/Tests/OutOfSchool.Tests.Common/TestDataGenerators/WorkshopRequiredPropertiesDtoGenerator.cs:
```diff
@@ -20,8 +20,7 @@ public static class WorkshopRequiredPropertiesDtoGenerator
         .RuleFor(w => w.Price, f => f.Random.Decimal())
-        .RuleFor(w => w.PayRate, f => f.PickRandom<PayRateType>())
-        .RuleFor(w => w.PreferentialTermsOfParticipation, f => f.Lorem.Paragraph())
+        .RuleFor(x => x.PayRate, f => f.PickRandom(PayRateType.Hour, PayRateType.Day, PayRateType.Month, PayRateType.Year, PayRateType.Course)).RuleFor(w => w.PreferentialTermsOfParticipation, f => f.Lorem.Paragraph())
```
Ok

HlibBondarev committed:
Fixed the problem mentioned by DmyMi and CodeRabbitAi.

HlibBondarev commented on OutOfSchool/OutOfSchool.WebApi.Tests/Validators/WorkshopV2DtoValidationTests.cs:
```diff
@@ -0,0 +1,396 @@
+        validDto.DateTimeRanges = [new() { StartTime = TimeSpan.Parse("18:00"), EndTime = TimeSpan.Parse("19:45"), Workdays = [DaysBitMask.Monday] }];
+    }
+
+    [Test]
+    public void Validate_WhenAvailableSeatsEqualsUintMaxValue_ShouldReturnNoValidationErrors()
+    {
+        // Arrange
+        var dto = validDto;
+        dto.AvailableSeats = uint.MaxValue;
+        var validationContext = new ValidationContext(dto);
+
+        // Act
+        var results = dto.Validate(validationContext).ToList();
+
+        // Assert
+        results.Should().BeEmpty();
+    }
```
Validation rules require that if Available Seats doesn't have limits it shoul be equels uint.MaxValue

HlibBondarev replied:
Ok

HlibBondarev commented on OutOfSchool/Tests/OutOfSchool.Tests.Common/TestDataGenerators/WorkshopRequiredPropertiesDtoGenerator.cs:
```diff
@@ -20,7 +20,7 @@ public static class WorkshopRequiredPropertiesDtoGenerator
         .RuleFor(w => w.ParentWorkshopId, f => null)
         .RuleFor(w => w.IsPaid, f => true)
         .RuleFor(w => w.Price, f => f.Random.Decimal())
```
Ok

HlibBondarev committed:
Fixed the problem mentioned by CodeRabbitAi.

HlibBondarev committed:
Added [MaxLength] attribute to the Keyword property WorkshopDescriptionDto class to limit the number of keywords.

DmyMi commented on OutOfSchool/OutOfSchool.BusinessLogic/Models/Workshops/TempSave/WorkshopContactsDto.cs:
```diff
@@ -1,10 +1,17 @@
 public class WorkshopContactsDto : WorkshopDescriptionDto, IHasContactsDto<Workshop>
 {
-    [ModelBinder(BinderType = typeof(JsonModelBinder))]
+    [CollectionNotEmpty(ErrorMessage = "At least one contact is required")]
```
`[ModelBinder(BinderType = typeof(JsonModelBinder))]` should not be removed! This field is sent as JSON when we use multipart form data. It can break our `WorkshopCreateRequestDto` in the V2 api.

HlibBondarev replied:
Ok

DmyMi commented on OutOfSchool/OutOfSchool.BusinessLogic/Models/Workshops/WorkshopBaseDto.cs:
```diff
@@ -102,13 +104,10 @@ public class WorkshopBaseDto : IValidatableObject, IHasContactsDto<Workshop>
     public string ProviderTitleEn { get; set; } = string.Empty;
 
-    [EnumDataType(typeof(ProviderLicenseStatus), ErrorMessage = Constants.EnumErrorMessage)]
     public ProviderLicenseStatus ProviderLicenseStatus { get; set; } = ProviderLicenseStatus.NotProvided;
```
`[EnumDataType(typeof(ProviderLicenseStatus), ErrorMessage = Constants.EnumErrorMessage)]`
is a very important validation and should not be removed for any Enum field.

HlibBondarev replied:
Ok

DmyMi commented on OutOfSchool/OutOfSchool.BusinessLogic/Models/CompetitiveEvent/CompetitiveEventDescriptionItemDto.cs:
```diff
@@ -1,20 +1,26 @@
     public Guid Id { get; set; }
 
-    [Required]
+    [Required(ErrorMessage = "Description heading is required")]
```
`(ErrorMessage = "Description heading is required")` is the default message anyway :) no need to add it, pure `[Required]` is enough

HlibBondarev replied:
Ok

DmyMi commented on OutOfSchool/OutOfSchool.BusinessLogic/Models/CompetitiveEvent/TempSave/CompetitiveEventDescriptionDto.cs:
```diff
@@ -37,17 +37,20 @@ public class CompetitiveEventDescriptionDto : CompetitiveEventAboutDto
     [MustContain(RequiredCharacterType.AnyLetter)]
     public string VenueName { get; set; }
 
+    [Required(ErrorMessage = "Information about the selection is required")]
```
here and everywhere the default "field is required" message can be omitted.

HlibBondarev replied:
Ok

DmyMi commented on OutOfSchool/OutOfSchool.BusinessLogic/Models/Workshops/WorkshopCreateRequestDto.cs:
```diff
@@ -13,6 +14,13 @@ public class WorkshopCreateRequestDto : WorkshopContactsDto
     
     public Guid? DefaultTeacherId { get; set; }
+
+    public override IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
```
do we really need to override it if the implementation is default from base class? question is for here and other DTOs where just `base` method is used

HlibBondarev replied:
No.
The keyword - "Override" is required. The code is used to validate all classes in the WorkshopDto hierarchy.

DmyMi commented on OutOfSchool/OutOfSchool.WebApi.Tests/OutOfSchool.WebApi.Tests.csproj:
```diff
@@ -31,6 +31,7 @@
 
   <ItemGroup>
     <ProjectReference Include="..\OutOfSchool.AuthCommon\OutOfSchool.AuthCommon.csproj" />
+    <ProjectReference Include="..\OutOfSchool.Common\OutOfSchool.Common.csproj" />
```
Common project is a transitive dependency of both WebAPi and AuthCommon, it may not be required to add it explicitly

HlibBondarev replied:
Ok

HlibBondarev committed:
Fixed the problem mentioned by DmyMi.

HlibBondarev committed:
Fix validation for the properties (RegistrationStartTime, RegistrationEndTime, ScheduledStartTime, ScheduledEndTime) of the CompetitiveEventAboutDto.

HlibBondarev commented on OutOfSchool/OutOfSchool.BusinessLogic/Models/Workshops/WorkshopBaseDto.cs:
```diff
@@ -16,19 +16,20 @@ public class WorkshopBaseDto : IValidatableObject, IHasContactsDto<Workshop>
-    [MaxLength(Constants.MaxWorkshopShortTitleLength,ErrorMessage = "This field must contain from 1 to 60 characters.")]
+    [Required]
+    [MinLength(Constants.MinWorkshopShortTitleLength, ErrorMessage = "This field must contain from 1 to 60 characters.")]
+    [MaxLength(Constants.MaxWorkshopShortTitleLength, ErrorMessage = "This field must contain from 1 to 60 characters.")]
```
Ok

HlibBondarev committed:
1) Removed the ErrorMessage parameter from the [MinLength] and [MaxLength] attributes of the ShortTitle property in the WorkshopBaseDto class.
2) Removed the ErrorMessage parameter from the [Required] attribute of the ShortTitle and Title properties in the WorkshopBaseDto class.

HlibBondarev committed:
Fix validation for the DateTimeRange property of the WorkshopBaseDto and WorkshopMainRequiredPropertiesDto classes.

HlibBondarev commented on OutOfSchool/OutOfSchool.BusinessLogic/Models/Workshops/WorkshopBaseDto.cs:
```diff
@@ -166,113 +168,103 @@ public class WorkshopBaseDto : IValidatableObject, IHasContactsDto<Workshop>
+                }
 
-            if (keywordsList.Any(string.IsNullOrWhiteSpace))
-            {
-                yield return new ValidationResult(
-                    "Keyword cannot be empty or whitespace.",
-                    new[] { nameof(Keywords) });
+                string trimmedKeyword = (keyword ?? string.Empty).Trim();
+                cleanedKeywords.Add(trimmedKeyword);
+                totalLength += trimmedKeyword.Length;
+
+                // Check 2: Single keyword max length
+                if (trimmedKeyword.Length > Constants.MaxLengthOfOneKeyword)
+                {
+                    yield return new ValidationResult(
+                        $"Keyword must be no longer than {Constants.MaxLengthOfOneKeyword} characters.",
+                        [nameof(Keywords)]);
+                }
```
Ok

HlibBondarev committed:
Improved validation for the Keywords property of the WorkshopBaseDto and WorkshopDescriptionDto classes.

HlibBondarev committed:
Added migration — 20251115101013_FixValidationForWorkshopAndWorkshopDescriptionItem.
